src/translate/compare.rs

```diff
@@ -0,0 +1,173 @@
+    }
+
+    pub fn to_string(&self) -> &'static str {
+        match self {
+            ComparisonOperator::Gt => ">",
+            ComparisonOperator::Ge => ">=",
+            ComparisonOperator::Lt => "<",
+            ComparisonOperator::Le => "<="
+        }
+    }
+}
+
+pub fn translate_lexical_comparison(
+    meta: &mut TranslateMetadata,
+    operator: ComparisonOperator,
+    left: &Expr,
+    right: &Expr
+) -> FragmentKind {
+    let left = {
+        let left_stmt = VarStmtFragment::new("__left_comp", left.get_type(), left.translate(meta));
+        let left_expr = VarExprFragment::from_stmt(&left_stmt).with_array_to_string(true).to_frag();
+        meta.stmt_queue.push_back(left_stmt.to_frag());
+        left_expr
+    };
+    let right = {
+        let right_stmt = VarStmtFragment::new("__right_comp", right.get_type(), right.translate(meta));
+        let right_expr = VarExprFragment::from_stmt(&right_stmt).with_array_to_string(true).to_frag();
+        meta.stmt_queue.push_back(right_stmt.to_frag());
+        right_expr
+    };
+    let (primary_operator, secondary_operator) = operator.get_bash_lexical_operators();
+    let expr = if let Some(secondary_operator) = secondary_operator {
+        fragments!("[[ ", left.clone(), primary_operator, right.clone(), " || ", left, secondary_operator, right, " ]] && echo 1 || echo 0")
+    } else {
+        fragments!("[[ ", left, primary_operator, right, " ]] && echo 1 || echo 0")
+    };
+    SubprocessFragment::new(expr).to_frag()
+}
+
+fn create_variable_length_getter(
+    meta: &mut TranslateMetadata,
+    name: &str,
+    expr: &Expr
+) -> VarExprFragment {
+    let var_stmt = VarStmtFragment::new(name, expr.get_type(), expr.translate(meta));
+    let var_expr = VarExprFragment::from_stmt(&var_stmt).with_length_getter(true);
+    meta.stmt_queue.push_back(var_stmt.to_frag());
+    var_expr
+}
+
+fn create_variable_with_smaller_number(
+    name: &str,
+    left: VarExprFragment,
+    right: VarExprFragment
+) -> (VarStmtFragment, VarExprFragment) {
+    let value = SubprocessFragment::new(
+        fragments!("(( ", left.clone().to_frag(), " < ", right.clone().to_frag(), "))",
+            " && echo ", left.clone().to_frag(),
+            " || echo ", right.clone().to_frag())).to_frag();
+    let len_stmt = VarStmtFragment::new(name, Type::Num, value);
+    let len_expr = VarExprFragment::from_stmt(&len_stmt).with_render_type(VarRenderType::NameOf);
+    (len_stmt, len_expr)
+}
+
+fn compare_array_lengths(
+    left_len: VarExprFragment,
+    right_len: VarExprFragment,
+    operator: ComparisonOperator
+) -> FragmentKind {
+    let (op, eq) = operator.get_bash_lexical_operators();
+    let comparison_fragment = fragments!(left_len.clone().to_frag(), op, right_len.clone().to_frag());
+    let full_comparison_fragment = if eq.is_some() {
+        fragments!("(( ", left_len.clone().to_frag(), " == ", right_len.clone().to_frag(), " || ", comparison_fragment, " ))")
+    } else {
```

Comment:
I don't like that you assume that the second operator will always be an equal sign and that you use it just a boolean value. You've created a fragment, why not use it?

Reply:
Right... the `eq` parameter will always be an equal sign. I'll change the code to make it more sense.